README.md
# mini-deluxe

Minimal web server for Bun.js.
mini-deluxe extends `Bun.serve()` and creates a simple wrapper with useful features such as handlers, handler modes, websocket support, and more.

This whole library is under 1000 lines and will always be.

## Specification

- Custom Handlers
- File-only mode
- ~~WebSockets~~ (Not currently supported...yet)
- Very easy to understand configuration
- Ease of use handler builders
- Handlers recieve input from PHP-like arguments

To install dependencies:

```bash
bun install
```

To run:

```bash
bun run index.ts
```

## Authors

- @tux7k <tux@nuk3.org>
- @datkat21 <kat@cherries.to>


This project was created using `bun init` in bun v1.0.0. [Bun](https://bun.sh) is a fast all-in-one JavaScript runtime.


CHANGELOG.md
# Changelog

## v1.0.1
- Added File Only Mode
- Fixed two types that were the same (``Route`` and ``ReqHandler``)
- Removed useless functions
- Added exports to main file, making it more usable in projects

## v1.0.2

- add redirects when page has no "/" at the end


examples/test_foh.ts
import {
  createCookie,
  editCookie,
  pageBuilder,
  rawFile,
} from "../handlers/handling";
import miniDeluxe from "../index";

function fofHandler() {
  return new Response(
    `<!DOCTYPE html><html><head><style>html,body{background-color:#1a1a1a;color:#ffffff;font-family:system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;padding: 20px;}</style></head><body><h1>404 Not Found</h1><p>The file you were looking for does not exist.</p></body></html>`,
    {
      status: 404,
      headers: {
        "Content-Type": "text/html",
      },
      statusText: "Not Found",
    }
  );
}

const m = miniDeluxe.miniServer({
  port: 3000,
  fofHandler,
  options: {
    fileOnlyMode: true,
    staticFilePath: "../public",
  },
});


examples/test_server.ts
import {
  RouteArguments,
  createCookie,
  editCookie,
  pageBuilder,
  rawFile,
} from "../handlers/handling";
import miniDeluxe from "../index";

const m = miniDeluxe.miniServer({
  port: 3000,
  // options: {
  //   fileOnlyMode: true
  // },
  handlers: [
    {
      urlPath: "/",
      handler: async function (data: RouteArguments) {
        return pageBuilder(
          "<h1>New Webs</h1><small><em>sponsored by cherries.to & tuck</em></small>"
        );
      },
    },
    {
      urlPath: "/test",
      handler: async function (data: RouteArguments) {
        return new Response(JSON.stringify(data.params), {
          headers: { "content-type": "application/json" },
        });
      },
    },
    {
      urlPath: "/turtle",
      handler: async function (data: RouteArguments) {
        return rawFile("../public/sample.png");
      },
    },
    {
      urlPath: "/cookie/make",
      handler: async function (data: RouteArguments) {
        return createCookie("cherries.to", "yes!");
      },
    },
    {
      urlPath: "/cookie/edit",
      handler: async function (data: RouteArguments) {
        return editCookie("cherries.to", "edited");
      },
    },
    {
      urlPath: "/error",
      handler: async function (data: RouteArguments) {
        throw new TypeError("Bad Error");
      },
    },
  ],
});


handlers/serverConstructor.ts
import { Server, WebSocketHandler } from "bun";
import {
  ReqHandler,
  fileOnlyReqHandler,
  generateRequestHandler,
  returnNotFound,
} from "./handling";

interface MiniDxServerOptions {
  enableWs?: boolean;
  staticFilePath?: string;
}

interface MiniDxServerConfig {
  port?: number;
  options?: MiniDxServerOptions;
  handlers?: ReqHandler[];
  fofHandler?: Function;
  errorHandler?: any;
}

export function miniServer(config: MiniDxServerConfig): Server {
  // Assign default port
  if (!config.port) config.port = 3000;
  // Assign default FoF handler
  if (!config.fofHandler) config.fofHandler = returnNotFound;
  // Assign default error handler
  if (!config.errorHandler)
    config.errorHandler = function (error: Error) {
      console.log(error);
      if (error.name === "ENOENT") {
        //@ts-ignore ts stuff :)
        return config.fofHandler();
      }

      return new Response(
        `<!DOCTYPE html><html><head><style>html,body{background-color:#1a1a1a;color:#ffffff;font-family:system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;padding: 20px;}</style></head><body><h1>${
          error.name
        } ${error.cause !== undefined ? `(${error.cause})` : ""}</h1><p>${
          error.message
        }</p><pre>${error.stack
          ?.replaceAll("<", "&lt;")
          .replaceAll(">", "&gt;")}</pre></body></html>`,
        {
          status: 500,
          headers: {
            "Content-Type": "text/html",
          },
          statusText: "Internal Server Error",
        }
      );
    };
  if (!config.options)
    config.options = {
      enableWs: false,
      staticFilePath: "public",
    };

  if (config.handlers === undefined) {
    /**
     * This text shows up somewhere
     */
    throw new Error("No handlers assigned to your server");
  } else {
    config.handlers = config.handlers.map((h) => {
      if (h.fomOptions?.enabled === true) {
        h.handler = fileOnlyReqHandler(h.urlPath, h.fomOptions.directory);
        return h;
      } else {
        return h;
      }
    });
  }
  if (!config.handlers) config.handlers = [];

  // Generate request handler
  const fetchHandler = generateRequestHandler(
    config.handlers,
    config.fofHandler
  );

  // Return final server
  return Bun.serve({
    port: config.port,
    fetch: fetchHandler,
    error: config.errorHandler,
  });
}


handlers/handling.ts
interface FileOnlyOptions {
  enabled: boolean;
  directory: string;
}

interface ReqHandler {
  urlPath: string;
  handler?: CallableFunction;
  fomOptions?: FileOnlyOptions;
}

export type { ReqHandler, FileOnlyOptions };

export function returnNotFound() {
  return new Response("404 Not Found", {
    status: 404,
    headers: { "Content-Type": "text/html" },
  });
}

export interface RouteArguments {
  req: Request;
  url: URL;
  route: ReqHandler;
  params: URLSearchParams;
}

export function generateRequestHandler(
  handlers: any[],
  fofHandler: Function = returnNotFound
) {
  async function requestHandler(req: Request): Promise<Response> {
    // Generate URL from request data
    const url = new URL(req.url);

    // Fall back to public directory
    let urlPath = url.pathname;

    // default handlers built into the program
    if (urlPath.endsWith("/")) {
      urlPath = "/index.html";
    }

    let route: ReqHandler | null;
    try {
      route = handlers.find((r: ReqHandler) => {
        if (r.fomOptions !== undefined && r.fomOptions.enabled === true) {
          // console.log("fom found:", r);
          if (url.pathname.startsWith(r.urlPath))
            // if (r.path.exec(url.pathname) !== null && r.method === req.method)
            return r;
        } else {
          if (r.urlPath.endsWith(url.pathname))
            // if (r.path.exec(url.pathname) !== null && r.method === req.method)
            return r;
        }
      });
    } catch (e: any) {
      return new Response(e, { status: 500 });
    }

    // If there's no such route, show a 404
    if (!route) return fofHandler();

    const params = new URLSearchParams(url.search);

    // Run the route's handler
    if (route.handler !== undefined)
      return await route.handler({ req, url, route, params } as RouteArguments);
    else return fofHandler();
  }

  return requestHandler;
}

import { join } from "path";

export function fileOnlyReqHandler(
  prefix: string = "/public",
  staticFilePath: string = "public"
) {
  async function requestHandler(req: Request) {
    // Generate URL from request data
    const url = new URL(req.url);

    // Fall back to public directory
    let urlPath = url.pathname.substring(prefix.length);

    // default handlers built into the program
    if (urlPath === "") {
      // redirect to /
      return new Response("Redirect", {
        status: 301,
        statusText: "Moved",
        headers: { Location: prefix + "/" },
      });
    }
    if (urlPath.endsWith("/")) {
      urlPath = "/index.html";
    }

    // console.log(prefix, staticFilePath, urlPath);

    return new Response(Bun.file(join(staticFilePath, urlPath)));
  }

  return requestHandler;
}

/**
 * Build a page and return with content type. Meant for simple API serverside uses like sqlite3 just to return extra comments.
 */
export function pageBuilder(html: string, contentType: string = "text/html") {
  return new Response(html, {
    headers: { "content-type": contentType },
  });
}

/// Create a cookie by ${name} and ${value}
export function createCookie(name: string, value: string) {
  return new Response("Cookie Created", {
    headers: { "Set-Cookie": `${name}=${value}` },
  });
}

/// Edit an already existing cookie by ${name} and ${value}
export function editCookie(name: string, value: string) {
  return new Response("Cookie Edited", {
    headers: { Cookie: `${name}=${value}` },
    status: 301,
    statusText: "Moved",
  });
}

/// Delete a cookie
export function deleteCookie(name: string, value: string) {
  return new Response("Cookie Edited", {
    headers: {
      "Set-Cookie": `${name}=${value}; path=/; expires=Thu, 01 Jan 1970 00:00:00 GMT`,
    },
    status: 301,
    statusText: "Moved",
  });
}

/// Return any raw file you would like, headers will be filled out by Bun.file()
export function rawFile(path: string) {
  // return rawFile = Bun.file -> Response;
  return new Response(Bun.file(path));
}
